src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -1327,40 +1326,93 @@ protected override void AllocAndTransformNativeToManaged(ILCodeStream codeStream
 
     class UnicodeStringMarshaller : Marshaller
     {
+        private bool ShouldBePinned
+        {
+            get
+            {
+                return MarshallerType != MarshallerType.Field
```

Comment:
This is missing a few cases. See our ilmarshallers.cpp:

bool ILWSTRMarshaler::CanUsePinnedManagedString(DWORD dwMarshalFlags)
{
    LIMITED_METHOD_CONTRACT;
    return IsCLRToNative(dwMarshalFlags) && !IsByref(dwMarshalFlags) && IsIn(dwMarshalFlags) && !IsOut(dwMarshalFlags);
}